<link rel="stylesheet" href="static/help/help.css">
<div class="col-md-12">
	<div class="panel panel-default">
		<div class="panel-heading">Help</div>
		<div class="panel-body">
			<h1>Introduction</h1>
			PartsGenie allows for the design and optimisation of novel,
			reusable synthetic biology parts.
			<p>This user interface consists of a number of components:
			<ul>
				<li><b>Parts</b>: an interactive drag-and-drop interface for
				assembling parts from features such as promoters, ribosome
				binding sites (RBS) and coding sequences (CDS).
				<li><b>Feature</b>: a panel allowing for the specification of
				parameters relating to individual features.
				<li><b>Filters</b>: a panel for specifying filters to consider
				in optimising the parts.
				<li><b>Results</b>: a panel for displaying and querying results
				of the part optimisations.
			</ul>
			<p>In addition, there is also the functionality for saving optimised
			parts to an instance of JBEI-ICE.
			
			<h1>Parts</h1>
			The <b>Parts</b> panel offers a simple drag-and-drop interface for
			defining multiple parts from a collection of features represented
			as buttons.
			<p><img src="static/help/parts.png" alt="Parts" class="help"/>
			<p>The user can build up a part by dragging-and-dropping features
			from the palette of grey buttons. Tablet and smartphone users can
			add features to the part by double-clicking on the feature button
			in the palette.
			<p>The features available are described below in the feature section.
			<p>Individual parts may be named.
			<p>Upon selecting a given feature, the <b>Feature</b> panel becomes
			visible, allowing the specification of parameters specific to that
			feature. Examples of this include an amino acid sequence for a CDS,
			or a translation initiation rate (TIR) for an RBS. The feature will
			be bordered by a red warning box until these necessary features are
			specified.
			<p>Note that a restriction exists such that a RBS <b>must</b> be
			immediately followed by a CDS. Again, the RBS will be bordered by
			a red warning box if this restriction is not adhered to.
			<p>In the case of optimising RBS or CDS, an <b>Organism</b> must be
			specified from the autofill field.
			<p>Multiple parts may be specified, by either clicking the '+' icon,
			or by duplicating a specified part through the Copy icon. Individual
			parts can be viewed and edited by clicking the numbered pagination
			panel. Designs may be deleted, as can individual features within parts.
			<p>Upon completing the set of parts, and adhering to all restrictions,
			the <b>Submit</b> button will turn green, and the job may be submitted.
			
			<h1>Feature</h1>
			The <b>Feature</b> panel is viewed on selecting a given feature in
			the <b>Design</b> panel.
			<p>The <b>Feature</b> panel is specific to the
			selected feature, which can be one of the following types:
			<ul>
				<li><b>Defined sequence</b>.
				<li><b>Randomised sequence</b>.
				<li><b>Sequence of defined melting temperature</b>.
				<li><b>Origin of replication</b>.
				<li><b>Promoter</b>.
				<li><b>Ribosome binding site</b>.
				<li><b>Coding sequence</b>.
				<li><b>Fixed coding sequence</b>.
				<li><b>Terminator</b>.
			</ul>
			<p>In each case, an individual <b>Feature</b> can be named.
			<p>In the cases of <b>Defined sequence</b>, <b>Origin of replication</b>,
			<b>Promoter</b> and <b>Terminator</b>, the only required parameter
			is the nucleotide <b>Sequence</b> itself. This sequence is fixed and
			not optimised by the PartsGenie algorithm.
			<p><b>Randomised sequence</b> requires a <b>Length</b> of the sequence,
			which is periodically mutated through the optimisation sequence to
			optimise the part as a whole.
			<p>A <b>Sequence of defined melting temperature</b> requires a defined
			melting temperature (<b>Tm</b>).
			<p><b>Ribosome binding site</b>s require a desired translation initiation
			rate <b>TIR</b>.
			<p><b>Coding sequence</b>s require an <b>Amino acid sequence</b> to
			be specified, which may be extracted automatically through the
			<b>Search Uniprot</b> facility:
			<p><img src="static/help/uniprot.png" alt="Search Uniprot" class="help bordered"/>
			
			<h1>Filters</h1>
			The <b>Filters</b> facility allow for the specification of constraints
			that must be applied during the optimisation process.
			A <b>Max repeating nucleotides</b> may be specified, preventing long
			runs of repeating nucleotides which impede synthesis of designed parts.
			<b>Excluded codons</b> and  <b>Excluded restriction sites</b> may
			be specified.
			<p>In all cases, such <b>Filters</b> should be kept to a minimum,
			as specifying too many unnecessary constraints will increase optimisation
			time and, in some cases, prevent optimisation altogether. For example,
			specifying either too few max repeating nucleotides, or too many excluded
			codons or restriction sites may result in the design space being constraint
			to such an extent that no valid reverse translation of an amino acid
			coding sequence exists.
			
			<h1>Results</h1>
			Upon completion of an optimisation job (including one or more parts),
			the <b>Results</b> panel is displayed.
			<p>The <b>Results</b> panel allows for the inspection of all optimised
			parts, allowing for the part as a whole, and individual features,
			to be viewed.
			<p>Individual features can be viewed by selecting the feature, and
			metadata specific to that feature can be inspected. (This may include
			<b>TIR</b> for a RBS, or Codon Adaptation Index (<b>CAI</b>) - a measure of how
			optimised codon usage is for the target host organism - for CDS features.)
			<p><img src="static/help/results.png" alt="Results" class="help"/>
			
			<h1>ICE</h1>
			Upon clicking the <b>ICE</b> button, the user is presented with a
			form for logging into an instance of
			<a href="http://ice.jbei.org" target="_out">JBEI ICE</a>.
			The required parameters
			are <b>URL</b> of the JBEI ICE instance, <b>Username</b> and <b>Password</b>.
			Finally, a <b>Group</b> may be specified, allowing any submitted designed
			part to be shared between members of a JBEI ICE defined group.
			<p>Once a user has logged into ICE, the <b>ICE</b> button will turn
			green, after which, parts may be <b>Save</b>d to JBEI ICE from the
			<b>Results</b> panel.
			
			<h1>Further details</h1>
			Further details are available from the publication describing this work.
			Please cite this reference in any scientific disclosure that utilises
			<b>PartsGenie</b>.
			<p><b>PartsGenie: an integrated tool for optimising and sharing synthetic biology parts</b>.
			<br/>Neil Swainston, Mark Dunstan, Adrian J Jervis, Christopher J
			Robinson, Pablo Carbonell, Alan R Williams, Jean-Loup Faulon, Nigel
			S Scrutton, Douglas B Kell.
			<br/><i>bioRxiv</i>, 189357
			<br/><a href="https://doi.org/10.1101/189357" target="_out">https://doi.org/10.1101/189357</a>
		</div>
	</div>
</div>